create broadcast performance improvements

schwers opened the pull request:
This patch is includes the fixes I've been using to address the performance overhead from unmounting incurred by the implmentation of `createBroadcast`. This should fix the issue: https://github.com/styled-components/styled-components/issues/931#issuecomment-318827123 (thanks for the ping! intended to open this PR this week).

It sounds like there's some talk of switching to `theming`, so I wanted to open this, and another branch of perf improvements I've been trying locally for feedback.

I tried using `brcast` directly and then had slightly better perf using a slightly different implementation; left some notes describing the difference in commit messages.

philpl commented:
Hiya, thanks for the PR :+1:

I think it'd be wiser to switch directly to `theming`. This was a longstanding plan and would achieve the same, perf-wise.

cc @mxstbr

mxstbr commented:
I'm happy to switch to `theming` but if this implementation is advantageous over `brcast` I'd rather merge this for noe until we can get a patch upstream?

/cc @vesparny

philpl commented:
@mxstbr the brcaat lib switched to this implementation a while ago :tada: With the tiny, nice detail that the unsubscribe function uses the id in a closure instead of ever passing it out: https://github.com/vesparny/brcast/blob/master/index.js

vesparny commented:
The implementation in this PR is slightly different because instead allocating the unsubscribe function it returns a subscription id.

I'd love to benchmark perf differences (how?) and change accordingly brcast implementation if 
Worth. 

That would be a breaking change in case.

vesparny commented:
I've already ditched the Array.forEach in favor of the more performant for loop. It has been released in brcast@2.0.2

mxstbr commented:
@philpl did you read the OP? :wink: 

> I tried using brcast directly and then had slightly better perf using a slightly different implementation; left some notes describing the difference in commit messages.

philpl commented:
@mxstbr yea, but compared to the previous impl it should already be faster, and we do want to switch to theming anyway. There is no need for a step inbetween when we can fix this problem in brcast and can stop having to worry about this.

Also we're entering a territory where more accurate benchmarks are needed. Unsubscriptions are a relatively rare operation. How fast does that have to be compared to what we're optimising in v3, style injection and generation/processing?

schwers commented:
@philpl I think making the switch to `theming` makes sense; however this is currently a big bottleneck in a large application I'm working in. For our use case (virtualized scrolling), this patch's perceived performance improvement is on par with the [batch-updates branch](https://github.com/styled-components/styled-components/tree/batch-updates) branch. Is there anything I can do to help with the switch to `theming`?

Absolutely agree that more accurate benchmarks are needed. Do you have any recommendations for tests that are trying to measure something like GC optimizations? My intuition here would be to measure times for creating large numbers of non-trivial styled component trees.

> Unsubscriptions are a relatively rare operation.

I'd like to add that this patch, if I'm understanding V8s GC strategy, also has an impact on component mount times as well. `unsubscribe` as implemented in `brcast` has to allocate a new `unsubscribe` callback on mount, potentially triggering GC. 

I. agree that unsubscription is fairly uncommon, but I think the two biggest use-cases are performance sensitive; e.g. unmounting pages in response to routing changes (noted in https://github.com/styled-components/styled-components/issues/931), and virtualized scrolling.

schwers commented on src/hoc/withTheme.js:
```diff
@@ -23,25 +22,32 @@ const wrapWithTheme = (Component: ReactClass<any>) => {
 
     static contextTypes = {
-      [CHANNEL]: PropTypes.func,
+      [CHANNEL]: CONTEXT_CHANNEL_SHAPE,
```
@mxstbr, @philpl , moving convo re: breaking contextTypes change. 

TBF, I think we've all unintentionally relied on an implementation detail at some point or another ;). What about a `CHANNEL_NEXT` context for now and a deprecation warning issued when using the `unsubscribe` on `CHANNEL`?

philpl replied:
That sounds optimal indeed :)

mxstbr replied:
Yeah, let's warn that we've changed some internals that'll break third-party libraries eventually and that they should upgrade, that sounds perfect.

philpl commented:
@schwers

> I think making the switch to theming makes sense; however this is currently a big bottleneck in a large application I'm working in. For our use case (virtualized scrolling), this patch's perceived performance improvement is on par with the batch-updates branch branch. 

Hm, I actually never thought about these cases. Indeed with virtualised scolling, unsubscriptions are frequent and not as rare as what I was thinking of 😓 

> is on par with the batch-updates branch branch. 

That is quite surprising, since there's no css injection operation during unsubscription, which makes the comparison of a benchmark with its performance purely based on the number of rules and new rules that are being injected 🤔  But that also means that you must be seeing a significant jump in perf

> I'd like to add that this patch, if I'm understanding V8s GC strategy, also has an impact on component mount times as well. unsubscribe as implemented in brcast has to allocate a new unsubscribe callback on mount, potentially triggering GC.

I wasn't even thinking about this as well 😅 That is probably a good explanation of why this is more significant than what I expected.

> Is there anything I can do to help with the switch to theming?

So reconsidering all the points we've discussed, I think it's fair to say that we should prefer merging this for now rather than switching over to `theming`. Relying on the same core libs is nice in the long run, but before v3 we won't even have any interoperability with the other component-based css-in-js libs, so it's not important right now.

So I think @mxstbr is right:

> I'm happy to switch to theming but if this implementation is advantageous over brcast I'd rather merge this for noe until we can get a patch upstream?

:+1:

mxstbr commented:
Yeah I watched the app @schwers is building and switching pages was super painful to watch, so I'd love to land this!

vesparny commented:
I'd be happy to implement this also in `brcast`.
This way also `glamorous` and any lib leveraging `theming` would benefit .

schwers commented:
@philpl @mxstbr updated with the deprecation notice! and added some perf comparisons to `theming` in https://github.com/styled-components/styled-components/pull/1051#issuecomment-319249833

probablyup commented on src/utils/create-broadcast.js:
```diff
@@ -7,29 +7,36 @@
+  const publish = (nextState: mixed) => {
+    state = nextState
+    const listenerKeys = Object.keys(listeners)
+    for (let i = 0; i < listenerKeys.length; i += 1) {
```
Could use a for-in here instead to drop the Object.keys call and having to track index

schwers replied:
Could you elaborate on this one? This seems to be the fastest approach via https://jsperf.com/object-keys-vs-for-in-with-closure/137. My one thought is it would be nice to not have to allocate the array of object keys, but on the other hand, publish'ing new themes seems like it wouldn't happen too often that it's a problem.

probablyup replied:
Depends heavily on the browser. For iOS safari, for-in wins by 10%. Probably more without the hasOwnProperty check. 

I'd probably code to make the slowest browsers faster vs targeting desktop ones.

schwers replied:
Interesting, I didn't realize the divide was that much on iOS. Sounds good.

probablyup commented on src/utils/create-broadcast.js:
```diff
@@ -7,29 +7,36 @@
-      }
-    },
+  const unsubscribe = unsubID => {
+    delete listeners[unsubID]
```
Does delete still cause a deopt in v8? Used to turn a fast object to a slow one.

schwers replied:
Good catch, my guess is yes, I'll try comparing.

mxstbr replied:
Yeah it does, literally just talked to a v8 engineer about this. Apparently

```
listener[unsubID] = undefined
```

doesn't deopt. /cc @fhinkel

probablyup replied:
That's true, but it leaves a hole in the array. Just need to add a truthy check to the iterator. Nbd.

mxstbr replied:
Yeah I guess. Dunno what effect that would have.

probablyup replied:
Minimal I'd expect since it's a simple Boolean guard. if `(x === undefined) { continue; }` should be fine and lightweight

philpl replied:
> That's true, but it leaves a hole in the array. Just need to add a truthy check to the iterator. Nbd.

`delete` causes a deopt of how the object is stored and processed, but setting it to `undefined` and handling it is expected and has been optimised. So taking the `undefined` branch wouldn't cause a huge perf penalty, I'd guess

probablyup commented on src/utils/create-broadcast.js:
```diff
@@ -7,29 +7,36 @@
-        listeners = listeners.filter(item => item !== listener)
-      }
-    },
+  const unsubscribe = unsubID => {
```
Maybe use a regular function instead of an arrow function - should be a tad faster

schwers replied:
This is being compiled down to a regular function :)

probablyup replied:
A bound one though. Possible perf penalty for lexical binding we don't need?

schwers replied:
oh I see

schwers commented:
Moved the changes from discussion in here. I tried benchmarking in comparison, seems marginally better, but doesn't seem like a big enough change to measure reliably with my current setup (local app)

schwers commented:
Clarification: I mean the changes in https://github.com/styled-components/styled-components/commit/b6b9eccf3f4036aeadbc59474e4f04b022d4a0d4 didn't have a noticeable impact in my test app. 

@philpl did you get a chance to look at https://github.com/styled-components/styled-components/pull/1051#issuecomment-319249833 ?

Wondering if I should close this in favor of `theming` for now?

philpl commented:
@schwers

>  I mean the changes in b6b9ecc didn't have a noticeable impact in my test app.

Stuff like `delete` and `bind` are still so fast despited deopts that it's not a big bottleneck during runtime that is noticeable.

Regarding the comment: https://github.com/styled-components/styled-components/pull/1051#issuecomment-319249833

It looks like the difference to `theming` is a lot bigger than what I expected, once again 😆 

The problems right now are:

- The bundle size is bigger than this patch (arguably a big deal since everyone's eyes are on it right now)
- I'd expect it to be slightly slower based on the optimiations we've discussed here (but we don't yet have benchmarks for `theming` that include other optimisations here)
- The GC looks like it's sawtoothing (@probablyup) but comparing the nodes (48,500 vs 16,200) looks to me like it just looks like it. Also the time spent in GC is shorter and if you can tell us whether FPS is stabler in this PR, then I'd say the performance is better on all metrics

@mxstbr It's really a question of what we're trying to get out of `theming`. We can still contribute our findings, but maybe we should merge this PR, and wait for them to fix these issues?

schwers commented:
> The GC looks like it's sawtoothing (@probablyup) but comparing the nodes (48,500 vs 16,200) looks to me like it just looks like it.

FWIW I think some of the sawtoothing is a bug in our app 😄 

> Also the time spent in GC is shorter and if you can tell us whether FPS is stabler in this PR, then I'd say the performance is better on all metrics

Yup, it does seem more stable with this patch. I notice less 'flicker' where row elements have yet to render, and overall seems a little smoother.

I'd be happy to help `theming` implement these changes, though I won't have much time until later this week.

philpl commented:
@schwers That sounds really good actually! :+1:

@mxstbr Should we try to get this merged and then look into whether we can add `theming` on top to compare the perf with the optimisations that are already included in this PR? If `theming` turns out to be significantly slower, we wouldn't have to merge it right away, but can point out some sources of worse perf and wait for them to fix it up.

schwers commented:
@philpl Did you mean #1048? :) What do you think of a slightly more incremental approach? i.e. would you be open to merging this patch first, then as changes land in `theming` we could have a baseline to compare against?

w.r.t the plan for: https://github.com/styled-components/styled-components/pull/1049 I'll try to split out the static styles first into a PR in the next day or so.

so I'm thinking it could really be a series of small patches
* static styles optimization
* faster classname generation ([src from #1049](https://github.com/styled-components/styled-components/pull/1049))

I'm less sure about these changes:
* flatten updates ([src from #1049](https://github.com/styled-components/styled-components/pull/1049))
* native array methods -> `for` (not a major concern for most apps, but could reduce some incremental allocations for use cases like virtualized-scrolling)

philpl commented:
@schwers I was thinking of merging #1049 first actually, but if we can merge this PR first that's fine to me as well, as #1049 still needs some time for splitting it up, etc. It really depends on @mxstbr. I'm not too opinionated about the order of changes, and after some thinking, I don't consider the trade-offs of the theming PR worth it for now. it'll need some improvements.

> flatten updates (src from #1049)

We should probably exclude them from the PR for now, and split it up into even more smaller incremental PRs. That would greatly increase my confidence in every separate change, since it's easier to see where our tests fall short, what introduces regressions, and which part of it improves perf significantly.

> native array methods -> for (not a major concern for most apps, but could reduce some incremental allocations for use cases like virtualized-scrolling)

That would benefit from a separate PR as well, since we could then see how significant it actually is on its own

mxstbr commented:
I don't have time to do a deep-dive on this, please go ahead and merge in the order you want to 😜

schwers commented:
@philpl @mxstbr Can we start with merging this patch then? I've squashed the commits down to help with merging (should I not do that for this project if you use the github merge functionality?).

mxstbr commented:
If @philpl is happy with this patch I say we go ahead and merge this 🛳

philpl commented on src/models/AbstractStyledComponent.js:
```diff
@@ -12,10 +12,19 @@ export default class AbstractStyledComponent extends Component {
+
+  unsubscribeId: number = -1
+
+  unsubscribeFromContext() {
```
This component is basically to be deleted, since we already have the "BaseStyledComponent":

https://github.com/styled-components/styled-components/pull/1048/files#diff-8a1d0988ee5ae37265b4647fa41def2aR41

I'm not sure whether it's still worth keeping around, even if this code is actually going to be shared with the native implementation as well. The extra class actually seemed to slow instance instantiation down for me in some of my older perf flamegraphs, so I think we should aim for deleting it?

schwers replied:
Makes sense, the `AbstractStyledComponent` class doesn't seem to provide much utility vs having an extra class in the hierarchy. imho we should try to but as much into the base class as possible.

philpl commented on src/hoc/withTheme.js:
```diff
@@ -24,24 +24,32 @@ const wrapWithTheme = (Component: ReactClass<any>) => {
-      if (!this.context[CHANNEL]) {
-        throw new Error('[withTheme] Please use ThemeProvider to be able to use withTheme')
+      const styledContext = this.context[CHANNEL_NEXT]
+      if (!styledContext) {
```
We could check for `!== undefined` and below in the unmount we could check for `unsubscribeId !== -1` instead? The latter doesn't make a big difference imho

philpl commented on src/models/ThemeProvider.js:
```diff
@@ -6,26 +6,40 @@ import isFunction from 'is-function'
+
+const warnChannelDeprecated = once(() => {
+  // eslint-disable-next-line no-console
+  console.error(`Warning: Usage of \`context.${CHANNEL}\` as a function is deprecated. It will be replaced with the object on \`.context.${CHANNEL_NEXT}\` in a future version.`)
```
:tada:

philpl commented:
@mxstbr @schwers just some minor nits I wanted to address before merging this (sorry 😅 )

schwers commented:
@philpl no worries, good ideas! Incorporated your suggestions

philpl commented:
LGTM! :tada: :smile:

Are you alright with the changes, @mxstbr?

probablyup commented:
Just ran the benchmarks and they aren't super flattering at the moment:

```
v2.1.2
======

Results for browser
                                   styled         cssta        
---------------------------------  -------------  -------------
simple component                   0.113±0.019ms  0.011±0.006ms
prop changes                       0.254±0.045ms  0.051±0.011ms
prop shorthands                    0.126±0.043ms  0.009±0.003ms
prop shorthands with prop changes  0.294±0.121ms  0.047±0.002ms


branch (rebased against master locally)
=======================================

Results for browser
                                   styled         cssta        
---------------------------------  -------------  -------------
simple component                   0.174±0.044ms  0.013±0.002ms
prop changes                       0.314±0.064ms  0.053±0.010ms
prop shorthands                    0.167±0.058ms  0.011±0.002ms
prop shorthands with prop changes  0.322±0.104ms  0.044±0.003ms
```

That being said, benchmarks rarely reflect real-world perf.

mxstbr commented:
@probablyup does that benchmark include unmounting? (doesn't look like it to me) That's what this PR improves, everything else is mostly noise.

probablyup commented:
@mxstbr doesn't. shouldn't be hard to add though. I'll make myself a note to submit a PR

mxstbr commented:
Thanks!

philpl commented:
@probablyup based on the changes and the error range of the benchmark, I'd say it's not significant. And as Max says, it doesn't include actually include in mounting. Also subscribing to theme changes is not included as well, if I'm not mistaken.

Generally the benchmark leaves a lot to be desired and hasn't been a great indicator in my experience :/

philpl commented on src/models/StyledComponent.js:
```diff
@@ -38,18 +37,38 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
+      [CONTEXT_KEY]: PropTypes.instanceOf(StyleSheet),
+    }
+
+    static propTypes = {
```
Sorry, I just caught some more things. Can you remove the `contextTypes` and `propTypes` static property from the `BaseStyledComponent` please? :wink: they'll only need to be present on each `StyledComponent`.

Also, let's not add the `propTypes`. It has worked out fine without them before fortunately, and it might break some things and output unwanted warnings for some people. Even if it's the right thing to do, it's something we should save up for the major bump, as people really become annoyed when new warnings surface (even if they're warranted)

schwers commented:
@philpl 😅 I added the propTypes because of linter errors from the `react/prop-types` rule. just added https://github.com/styled-components/styled-components/pull/1048/commits/83ded61434979c5acb7bbdd5df7df6566377350b to remove that and worked around it with disable comments

philpl commented:
@schwers Weird! When did this rule become active? It never reported anything in this file before, I think

schwers commented:
@philpl it started happening when I removed the abstract base class. Seemed like it was previously no-op'ing instead of trying to resolve the parent propTypes

philpl commented:
@schwers OH! It didn't work anymore because we were extending the abstract class! It probably checks whether it's extending `Component`. Ok, anyway. Let's merge this for now and deal with that rule later
